pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * @see FixedByteValueReaderWriter
+ */
+public class VarLengthBytesValueReaderWriter implements Closeable, ValueReader {
+
+  /**
+   * Header used to identify the dictionary files written in variable length bytes
+   * format. Keeping this a mix of alphanumeric with special character to avoid
+   * collisions with the regular int/string dictionary values written in fixed size
+   * format.
+   */
+  private static final byte[] MAGIC_HEADER = StringUtil.encodeUtf8("....vl1;");
+  private static final int MAGIC_HEADER_LENGTH = MAGIC_HEADER.length;
+  private static final int INDEX_ARRAY_START_OFFSET = MAGIC_HEADER.length + Integer.BYTES;
+
+  private final PinotDataBuffer _dataBuffer;
+  private transient int _numElements;
+
+  public VarLengthBytesValueReaderWriter(PinotDataBuffer dataBuffer) {
+    this._dataBuffer = dataBuffer;
+  }
+
+  public static long getRequiredSize(byte[][] byteArrays) {
+    // First include the magic header, length field and offsets array.
+    long length = INDEX_ARRAY_START_OFFSET + Integer.BYTES * (byteArrays.length);
+
+    for (byte[] array: byteArrays) {
+      length += array.length;
+    }
+    return length;
+  }
+
+  public static boolean isVarLengthBytesDictBuffer(PinotDataBuffer buffer) {
+    // If the buffer is smaller than header + numElements size, it's not var length dictionary.
+    if (buffer.size() > INDEX_ARRAY_START_OFFSET) {
+      byte[] header = new byte[MAGIC_HEADER_LENGTH];
+      buffer.copyTo(0, header, 0, MAGIC_HEADER_LENGTH);
+
+      if (Arrays.equals(MAGIC_HEADER, header)) {
+        // Also verify that there is a valid numElements value.
+        int numElements = buffer.getInt(MAGIC_HEADER_LENGTH);
+        return numElements > 0;
+      }
+    }
+
+    return false;
+  }
+
+  private void writeHeader() {
+    for (int offset = 0; offset < MAGIC_HEADER_LENGTH; offset++) {
+      _dataBuffer.putByte(offset, MAGIC_HEADER[offset]);
+    }
+  }
+
+  public void init(byte[][] byteArrays) {
+    this._numElements = byteArrays.length;
+
+    writeHeader();
+
+    // Add the number of elements as the first field in the buffer.
+    _dataBuffer.putInt(MAGIC_HEADER_LENGTH, byteArrays.length);
+
+    // Then write the offset of each of the byte array in the data buffer.
+    int nextOffset = INDEX_ARRAY_START_OFFSET;
+    int nextArrayStartOffset = INDEX_ARRAY_START_OFFSET + Integer.BYTES * byteArrays.length;
+    for (byte[] array: byteArrays) {
+      _dataBuffer.putInt(nextOffset, nextArrayStartOffset);
+      nextOffset += Integer.BYTES;
+      nextArrayStartOffset += array.length;
+    }
+
+    // Finally write the byte arrays.
+    nextArrayStartOffset = INDEX_ARRAY_START_OFFSET + Integer.BYTES * byteArrays.length;
+    for (byte[] array: byteArrays) {
+      _dataBuffer.readFrom(nextArrayStartOffset, array);
+      nextArrayStartOffset += array.length;
+    }
+  }
+
+  int getNumElements() {
+    // Lazily initialize the numElements.
+    if (_numElements == 0) {
```

Comment:
initialize it to -1 to handle the case of an empty segment.

Reply:
Why not use a structure where byte arrays are stored at one and the offsets at the other? I agree that is useful for mutability, but with a pre-calculated length, mutable can be same as immutable

Reply:
Initializing this to `-1` now.
@mcvsubbu we can do that too but I'm not see any huge differences because of that. Can you think of any other advantages with that approach?